src/libguac/wol.c

```diff
@@ -0,0 +1,128 @@
+ * and place it in the character array.
+ * 
+ * @param packet
+ *     The character array that will contain the generated packet.
+ * 
+ * @param mac_address
+ *     The unsigned int representation of the MAC address to place in the packet.
+ */
+static void __guac_wol_create_magic_packet(unsigned char packet[],
+        unsigned int mac_address[]) {
+    
+    int i;
+    unsigned char mac[6];
+    
+    /* Concurrently fill the first part of the packet with 0xFF, and copy the
+     MAC address from the int array to the char array. */
+    for (i = 0; i < 6; i++) {
+        packet[i] = 0xFF;
+        mac[i] = mac_address[i];
+    }
+    
+    /* Copy the MAC address contents into the char array that is storing
+     the rest of the packet. */
+    for (i = 1; i <= 16; i++) {
+        memcpy(&packet[i * 6], &mac, 6 * sizeof(unsigned char));
+    }
+    
+}
+
+/**
+ * Send the magic Wake-on-LAN (WoL) packet to the specified broadcast address,
+ * returning the number of bytes sent, or zero if any error occurred and nothing
+ * was sent.
+ * 
+ * @param broadcast_addr
+ *     The broadcast address to which to send the magic WoL packet.
+ * 
+ * @param packet
+ *     The magic WoL packet to send.
+ * 
+ * @return 
+ *     The number of bytes sent, or zero if nothing could be sent.
+ */
+static ssize_t __guac_wol_send_packet(const char* broadcast_addr,
+        unsigned char packet[]) {
+    
+    int wol_bcast = 1;
+    struct sockaddr_in wol_dest;
+    int wol_socket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    
+    /* Attempt to set broadcast; exit with error if this fails. */
+    if (setsockopt(wol_socket, SOL_SOCKET, SO_BROADCAST, &wol_bcast, sizeof(wol_bcast)) < 0) {
+        return 0;
+    }
+    
+    /* Set server endpoint to the broadcast address. */
+    wol_dest.sin_family = AF_INET;
+    wol_dest.sin_port = htons(9);
+    wol_dest.sin_addr.s_addr = inet_addr(broadcast_addr);
+    
+    /* Send the packet and return number of bytes sent. */
+    return sendto(wol_socket, &packet, sizeof(unsigned char) * 102, 0,
+            (struct sockaddr*) &wol_dest, sizeof(wol_dest));
+ 
+}
+
+int guac_wol_wake(const char* mac_addr, const char* broadcast_addr,
+        const int wait_time) {
+    
+    unsigned char wol_packet[102];
+    unsigned int dest_mac[6];
+    
+    /* Parse mac address and return with error if parsing fails. */
+    if (sscanf(mac_addr, "%x:%x:%x:%x:%x:%x",
+            &(dest_mac[0]), &(dest_mac[1]), &(dest_mac[2]),
+            &(dest_mac[3]), &(dest_mac[4]), &(dest_mac[5])) != 6) {
+        return -1;
+    }
+    
+    /* Generate the magic packet. */
+    __guac_wol_create_magic_packet(wol_packet, dest_mac);
+    
+    /* Send the packet and record bytes sent. */
+    int bytes_sent = __guac_wol_send_packet(broadcast_addr, wol_packet);
+    
+    /* Sleep for the specified amount of time to wait for the remote
+     host to boot. */
+    sleep(wait_time);
```

Comment:
I'm happy to move the sleep to the caller at the protocol level - I have no argument with that.  i was trying to avoid as much code duplication as possible, but that's fine with me.  I'll move it out to the individual protocols and use the function you mentioned, which will probably also take care of the other issue I posted above with the connection timing out while sleeping.  Thanks!